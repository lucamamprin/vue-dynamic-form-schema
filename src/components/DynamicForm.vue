<template>
  <div>
    <component
        v-for="(field, key) in schema"
        :key="key"
        :name="key"
        :is="getComponent(field.type)"
        v-bind="{...field}"
        :id="generateUniqueIdFromNameSpace(field, key)"
        :described-by="generateUniqueIdFromNameSpace(field, key, true)"
        :value="getValue(field.type, value, key)"
        :min="validationParam(field.validations.minValue, 0)"
        :max="validationParam(field.validations.maxValue, null)"
        :minLength="validationParam(field.validations.minLength, 0)"
        :maxLength="validationParam(field.validations.maxLength, null)"
        :uppercase="field.uppercase"
        :autocomplete="field.autocomplete"
        @input="update(key, $event)"
        :required="getIsRequired(field)"
        :label="getLabel(field)"
        :invalid="$v.form[key].$anyError"
        :hasError="$v.form[key].$error && showErrors"
        :error="errorMessages[key]"
        :show-character-count="field.showCharacterCount"
        :dark-mode="darkMode"
        :reset-margin-top="firstFormElement === key"
    >
    </component>
  </div>
</template>

<script>
import {validationRules} from '../parsers/ValidationParser'

const InputGeneric = () => import('./FormElements/InputGeneric')
const InputTextarea = () => import('./FormElements/InputTextarea')
const InputSelect = () => import('./FormElements/InputSelect')
const InputRadioGroup = () => import('./FormElements/InputRadioGroup')
const InputCheckbox = () => import('./FormElements/InputCheckbox')

export default {
  name: 'DynamicForm',
  props: {
    schema: {type: Object, required: true},
    value: {type: Object, required: true},
    nameSpace: {
      type: String,
      required: true,
    },
    // by default show errors while the user is typing
    // if it's needed to show on submit, its value must be handled by the parent component
    showErrors: {
      type: Boolean,
      default: true,
    },
    darkMode: {
      type: Boolean,
      default: false,
    },
    initialValue: {
      type: [String, Boolean],
      default: null,
    },
  },
  data() {
    return {
      form: {},
    }
  },
  components: {
    InputTextarea,
    InputGeneric,
    InputSelect,
    InputRadioGroup,
    InputCheckbox,
  },
  computed: {
    firstFormElement() {
      return Object.keys(this.schema)[0]
    },
    errorMessages() {
      const validations = this.$v.form

      return Object.keys(this.schema).reduce((messages, key) => {
        const rules = this.schema[key].validations
        const rulesKeys = Object.keys(this.schema[key].validations)
        const validator = validations[key]

        if (!validator) return messages

        for (const rule of rulesKeys) {
          if (validator[rule] !== false) continue

          const message = rules[rule].message
          const params = rules[rule].params

          messages[key] = message.replace('{' + rule + '}', params)

          if (rule === 'isBetween') {
            messages[key] = message
                .replace('{min}', params[0])
                .replace('{max}', params[1])
          }

          return messages
        }

        return messages
      }, {})
    },
  },
  methods: {
    getComponent(type) {
      let component
      switch (type) {
        case 'textarea':
          component = 'InputTextarea'
          break
        case 'select':
          component = 'InputSelect'
          break
        case 'radio_group':
          component = 'InputRadioGroup'
          break
        case 'checkbox':
          component = 'InputCheckbox'
          break
        default:
          component = 'InputGeneric'
      }

      return component
    },
    getIsRequired(field) {
      return field.validations.hasOwnProperty('required')
          || field.validations.hasOwnProperty('isChecked')
    },
    generateUniqueIdFromNameSpace(field, name, isError = false) {
      return isError ? `${this.nameSpace}-${name}-error` : `${this.nameSpace}-${name}`
    },
    getLabel(field) {
      // add an asterisk in case of required field
      // wrap the asterisk with a span of aria-hidden=true
      // this prevents screen readers from reading "star" or "asterisk"
      // all screen readers will announce if a field is required thanks to the other attrs

      return this.getIsRequired(field) ?
          `${field.label}<span aria-hidden="true">*</span>`
          : field.label
    },
    getValue(type, value, key) {
      // checkboxes handle values differently from the other inputs
      return type === 'checkbox' ? this.schema[key].value : value[key]
    },
    getShowCharacterCount(field) {
      return typeof field.showCharacterCount === 'undefined' ? true : field.showCharacterCount
    },
    validationParam(prop, defaultValue) {
      return typeof prop !== 'undefined' ? prop.params : defaultValue
    },
    update(key, value) {
      this.form[key] = value
      this.$v.form[key].$touch()

      this.$emit('input', {
        ...this.value,
        [key]: value,
      })
    },
    // used externally
    validate() {
      this.$v.$touch()
      this.announceStatus()
    },
    // used externally
    getAllErrors() {
      return this.errorMessages
    },
    announceStatus() {
      this.$emit('status', {
        invalid: this.$v.$invalid,
      })
    },
  },
  validations() {
    return {form: validationRules(this.schema)}
  },
  created() {
    for (const name in this.schema) {
      this.$set(this.form, name, this.initialValue)
    }
  },
  mounted() {
    this.$emit('mounted', this.form)
  },
}
</script>
