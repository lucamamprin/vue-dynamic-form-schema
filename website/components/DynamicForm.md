# DynamicForm

## Props

<!-- @vuese:DynamicForm:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|schema|the JSON schema representing the form structure.|`Object`|`true`|-|
|value|-|`Object`|`true`|-|
|nameSpace|A unique string that works as ID for the form. This is useful in case you need two identical forms that have different data (eg: shipping and billing address).|`String`|`true`|-|
|showErrors|by default show errors while the user is typing if it's needed to show on submit, its value must be handled by the parent component|`Boolean`|`false`|true|
|darkMode|If set to `true`, add classes to handle a form in a dark background.|`Boolean`|`false`|false|
|initialValue|Set the initial value for each field. For checkboxes it is recommended setting this to `false`|`String` /  `Boolean`|`false`|null|

<!-- @vuese:DynamicForm:props:end -->


## Events

<!-- @vuese:DynamicForm:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|input|-|-|
|status|-|-|
|mounted|-|-|

<!-- @vuese:DynamicForm:events:end -->


