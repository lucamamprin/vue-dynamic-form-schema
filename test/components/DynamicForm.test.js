import Vue from "vue"
import { shallowMount } from "@vue/test-utils"
import DynamicForm from "../../src/components/DynamicForm"
import firstName from "./fixtures/fields/firstName.json"
import lastName from "./fixtures/fields/lastName.json"
import email from "./fixtures/fields/email.json"
import passwordRepeat from "./fixtures/fields/passwordRepeat.json"

const newUserSchema = {
  ...firstName,
  ...lastName,
  ...email,
  ...passwordRepeat,
}

let wrapper

describe("form", () => {
  beforeEach(() => {
    wrapper = shallowMount(DynamicForm, {
      propsData: {
        schema: newUserSchema,
        nameSpace: "new-user",
      },
    })
  })

  test("mounts correctly", async () => {
    await Vue.nextTick()
    expect(wrapper).toMatchSnapshot()
  })
})
