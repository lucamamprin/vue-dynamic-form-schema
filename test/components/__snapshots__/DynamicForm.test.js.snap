// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form mounts correctly 1`] = `
<div>
  <div class="input floating-label">
    <div class="position-relative"><input type="text" aria-describedby="new-user-firstName-error " name="firstName" id="new-user-firstName" min="0" minlength="0" placeholder=" " aria-labelledby="label-firstName" aria-invalid="false" aria-required="true" autocomplete="false" error="Name is required." validations="[object Object]" class=""> <label id="label-firstName" for="new-user-firstName">Name<span aria-hidden="true">*</span></label></div>
    <!---->
    <!---->
  </div>
  <div class="input floating-label">
    <div class="position-relative"><input type="text" aria-describedby="new-user-lastName-error " name="lastName" id="new-user-lastName" min="0" minlength="0" placeholder=" " aria-labelledby="label-lastName" aria-invalid="false" aria-required="true" error="Your last name is required." validations="[object Object]" class=""> <label id="label-lastName" for="new-user-lastName">Last name<span aria-hidden="true">*</span></label></div>
    <!---->
    <!---->
  </div>
  <div class="input floating-label">
    <div class="position-relative"><input type="email" aria-describedby="new-user-email-error " name="email" id="new-user-email" min="0" minlength="0" placeholder=" " aria-labelledby="label-email" aria-invalid="false" aria-required="true" autocomplete="off" error="This field is required" validations="[object Object]" class=""> <label id="label-email" for="new-user-email">Email<span aria-hidden="true">*</span></label></div>
    <!---->
    <!---->
  </div>
  <div class="input floating-label">
    <div class="position-relative"><input type="password" aria-describedby="new-user-password-error new-user-password-counter" name="password" id="new-user-password" min="0" minlength="0" maxlength="20" placeholder=" " aria-labelledby="label-password" aria-invalid="false" aria-required="true" autocomplete="off" error="The password is required." validations="[object Object]" class=""> <label id="label-password" for="new-user-password">Password<span aria-hidden="true">*</span></label></div>
    <div>
      <!---->
      <div class="input-character-counter">
        <div class="text-right text-light"><span id="new-user-password-counter" class="sr-only">
    0 characters on 20 available for Password&lt;span aria-hidden="true"&gt;&lt;/span&gt;
  </span>
          <p class="m-0">
            0 / 20
          </p>
        </div>
      </div>
    </div>
    <!---->
  </div>
  <div class="input floating-label">
    <div class="position-relative"><input type="password" aria-describedby="new-user-repeatPassword-error new-user-repeatPassword-counter" name="repeatPassword" id="new-user-repeatPassword" min="0" minlength="0" maxlength="20" placeholder=" " aria-labelledby="label-repeatPassword" aria-invalid="false" aria-required="true" autocomplete="off" error="This field is required." validations="[object Object]" class=""> <label id="label-repeatPassword" for="new-user-repeatPassword">Ripeti password<span aria-hidden="true">*</span></label></div>
    <div>
      <!---->
      <div class="input-character-counter">
        <div class="text-right text-light"><span id="new-user-repeatPassword-counter" class="sr-only">
    0 characters on 20 available for Ripeti password&lt;span aria-hidden="true"&gt;&lt;/span&gt;
  </span>
          <p class="m-0">
            0 / 20
          </p>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
